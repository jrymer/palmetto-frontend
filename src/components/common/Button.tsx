import React from 'react';
import styled from 'styled-components';

const Container = styled.div<ContainerProps>`
  margin: ${(props) => props.theme.margin};
  background-color: ${(props) =>
    props.disabled ? props.theme.colors.offsetGrey : props.theme.colors.primary};
  display: flex;
  align-items: center;
  justify-content: center;
  cursor: ${(props) => (props.disabled ? 'auto' : 'pointer')};
  width: ${(props) => props.width};

  &:hover {
    background-color: ${(props) =>
      props.disabled ? props.theme.colors.offsetGrey : props.theme.colors.primaryContrast};
  }

  &:active {
    background-color: ${(props) =>
      props.disabled ? props.theme.colors.offsetGrey : props.theme.colors.primaryContrast};
    transform: ${(props) => (props.disabled ? 'none' : 'scale(0.98)')};
  }
`;

const TextContainer = styled.span`
  text-align: center;
  font-size: var(--fs-500);
  padding: 1em 3em;
  color: #fff;
`;

interface ContainerProps {
  disabled: boolean;
  width: string;
}

interface ButtonProps {
  buttonText: string;
  disabled?: boolean;
  onSearch: () => void;
  width?: string;
}

/**
 * Renders a button
 *
 * @param {string} buttonText Text to display on the button
 * @param {boolean} disabled If the button is clickable or not
 * @param {() => void} onSearch Fire the weather API search
 * @param {string} width Optional width override
 * @return A reusable button component
 */
const Button: React.FC<ButtonProps> = ({
  buttonText,
  disabled = false,
  onSearch,
  width = 'var(--col-width)',
}) => {
  const handleClick = () => {
    if (!disabled) {
      onSearch();
    }
  };
  return (
    <Container disabled={disabled} onClick={handleClick} role="button" width={width}>
      <TextContainer>{buttonText}</TextContainer>
    </Container>
  );
};

export default Button;
