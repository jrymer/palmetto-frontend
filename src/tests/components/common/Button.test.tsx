import React from 'react';

import Button from '../../../components/common/Button';
import { createSnapshot, fireEvent, render, screen } from '../../testWrapper';

const mockOnSearch = jest.fn();

describe('Button tests', () => {
  const buttonText = 'Search';
  it('Button snapshot is valid with base props', () => {
    const snapshot = createSnapshot(<Button buttonText={buttonText} onSearch={mockOnSearch} />);
    expect(snapshot).toMatchSnapshot();
  });
  it('Can change button text', () => {
    render(<Button buttonText={buttonText} onSearch={mockOnSearch} />);
    expect(screen.getByText(buttonText)).toBeInTheDocument();
    render(<Button buttonText={'something else'} onSearch={mockOnSearch} />);
    expect(screen.getByText('something else')).toBeInTheDocument();
  });
  it('Can be clicked', () => {
    render(<Button buttonText={buttonText} onSearch={mockOnSearch} />);
    fireEvent.click(screen.getByRole('button'));
    expect(mockOnSearch).toHaveBeenCalled();
  });
  it('Cannot be clicked if disabled', () => {
    render(<Button buttonText={buttonText} disabled={true} onSearch={mockOnSearch} />);
    fireEvent.click(screen.getByRole('button'));
    expect(mockOnSearch).toHaveBeenCalledTimes(0);
  });
});
