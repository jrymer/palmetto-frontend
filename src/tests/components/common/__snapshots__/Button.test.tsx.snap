// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button tests Button snapshot is valid with base props 1`] = `
<div
  className="sc-bczRLJ fKnbXX"
  disabled={false}
  onClick={[Function]}
  role="button"
  width="var(--col-width)"
>
  <span
    className="sc-gsnTZi bBpxKd"
  >
    Search
  </span>
</div>
`;
